rclcpp/test/rclcpp/test_time.cpp

```diff
@@ -447,3 +449,134 @@ TEST_F(TestTime, test_overflow_underflow_throws) {
     test_time = rclcpp::Duration::from_nanoseconds(INT64_MIN) + rclcpp::Time(-1),
     std::underflow_error("addition leads to int64_t underflow"));
 }
+
+class TestClockSleep : public ::testing::Test
+{
+protected:
+  void SetUp()
+  {
+    // Shutdown in case there was a dangling global context from other test fixtures
+    rclcpp::shutdown();
+    rclcpp::init(0, nullptr);
+    node = std::make_shared<rclcpp::Node>("clock_sleep_node");
+    param_client = std::make_shared<rclcpp::SyncParametersClient>(node);
+    ASSERT_TRUE(param_client->wait_for_service(5s));
+  }
+
+  void TearDown()
+  {
+    node.reset();
+    rclcpp::shutdown();
+  }
+
+  rclcpp::Node::SharedPtr node;
+  rclcpp::SyncParametersClient::SharedPtr param_client;
+};
+
+TEST_F(TestClockSleep, sleep_until_basic_system) {
+  static const auto MILLION = 1000L * 1000L;
+  const auto milliseconds = 300;
+  rclcpp::Clock clock(RCL_SYSTEM_TIME);
+  auto delay = rclcpp::Duration(0, milliseconds * MILLION);
+  auto sleep_until = clock.now() + delay;
+
+  auto start = std::chrono::system_clock::now();
+  ASSERT_TRUE(clock.sleep_until(sleep_until));
+  auto end = std::chrono::system_clock::now();
+
+  EXPECT_GE(clock.now(), sleep_until);
+  EXPECT_GE(end - start, std::chrono::milliseconds(milliseconds));
+}
+
+TEST_F(TestClockSleep, sleep_until_basic_steady) {
+  static const auto MILLION = 1000L * 1000L;
+  const auto milliseconds = 100;
+  rclcpp::Clock clock(RCL_STEADY_TIME);
+  auto delay = rclcpp::Duration(0, milliseconds * MILLION);
+  auto sleep_until = clock.now() + delay;
+
+  auto steady_start = std::chrono::steady_clock::now();
+  ASSERT_TRUE(clock.sleep_until(sleep_until));
+  auto steady_end = std::chrono::steady_clock::now();
+
+  EXPECT_GE(clock.now(), sleep_until);
+  EXPECT_GE(steady_end - steady_start, std::chrono::milliseconds(milliseconds));
+}
+
+TEST_F(TestClockSleep, sleep_until_ros_time_enable_interrupt)
+{
+  auto clock = std::make_shared<rclcpp::Clock>(RCL_ROS_TIME);
+  rclcpp::TimeSource time_source;
+  time_source.attachNode(node);
+  time_source.attachClock(clock);
+
+  // 5 second timeout, but it should be interrupted right away
+  const auto until = clock->now() + rclcpp::Duration(5, 0);
+
+  // Try sleeping with ROS time off, then turn it on to interrupt
+  bool sleep_succeeded = true;
+  auto sleep_thread = std::thread(
+    [clock, until, &sleep_succeeded]() {
+      sleep_succeeded = clock->sleep_until(until);
+    });
+  // yield execution long enough to let the sleep thread get to waiting on the condition variable
+  std::this_thread::sleep_for(std::chrono::milliseconds(10));
```

Comment:
This feels like it isn't going to be long enough to not be flaky, particularly on Windows.  I would suggest we set this much higher, maybe something like 200 milliseconds.